dht_net.go

```diff
@@ -266,49 +288,166 @@ func (ms *messageSender) SendRequest(ctx context.Context, pmes *pb.Message) (*pb
 			}
 		}
 
-		mes := new(pb.Message)
-		if err := ms.ctxReadMsg(ctx, mes); err != nil {
-			ms.s.Reset()
-			ms.s = nil
+		resch := make(chan requestResult, 1)
+		select {
+		case ms.rch <- resch:
+		default:
+			// pipeline stall, log it and time it
+			evt := log.EventBegin(ctx, "dhtSendRequestStall", ms.dht.self, ms.p, pmes)
+			select {
+			case ms.rch <- resch:
+				evt.Done()
+			case <-ctx.Done():
+				evt.Done()
+				ms.lk.Unlock()
+				return nil, ctx.Err()
+			case <-ms.dht.ctx.Done():
+				evt.Done()
+				ms.lk.Unlock()
+				return nil, ms.dht.ctx.Err()
+			}
+		}
+
+		rcount := ms.rcount
+
+		ms.lk.Unlock()
+
+		rctx, cancel := context.WithTimeout(ctx, dhtReadMessageTimeout)
+		defer cancel()
+
+		var res requestResult
+		select {
+		case res = <-resch:
+
+		case <-rctx.Done():
+			return nil, rctx.Err()
```

Comment:
If the context is cancelled, do we want to kill off the stream? or is that handled elsewhere?

Reply:
This is the specific request context, so it's inappropriate to kill for the entire stream.

Reply:
Thinking a bit more about this, I think we do want to kill the whole stream after all.
The issue is that if we start having slow responses (over 1m), the pipeline will fill with everything timing out and become unusable until the responses are received (regardless of whether we have stopped waiting).

Reply:
Will implement with a "reset" directive to the message receiver.